app/code/Magento/Quote/Model/QuoteRepository.php

```diff
@@ -203,13 +215,15 @@ public function delete(\Magento\Quote\Api\Data\CartInterface $quote)
      * @param int $identifier
      * @param int[] $sharedStoreIds
      * @throws NoSuchEntityException
-     * @return Quote
+     * @return CartInterface
      */
     protected function loadQuote($loadMethod, $loadField, $identifier, array $sharedStoreIds = [])
     {
-        /** @var Quote $quote */
-        $quote = $this->quoteFactory->create();
-        if ($sharedStoreIds) {
+        /** @var CartInterface $quote */
+        $quote = $this->cartFactory->create();
+        if ($sharedStoreIds &&
+            ($quote instanceof Quote || method_exists($quote, 'setSharedStoreIds'))
```

Comment:
Remove first part of condition. `method_exists` is sufficient.

Reply:
@orlangur done